common/include/pcl/experimental/executor/omp_executor.hpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception, Inc.
+ *  Author: Shrijit Singh <shrijitsingh99@gmail.com>
+ *
+ */
+
+#pragma once
+
+#ifdef _OPENMP
+  #include <omp.h>
+#endif
+
+#include <pcl/experimental/executor/property.h>
+#include <pcl/experimental/executor/type_trait.h>
+
+namespace executor {
+
+template <typename Blocking, typename ProtoAllocator>
+struct omp_executor;
+
+#ifdef _OPENMP
+template <>
+struct is_executor_available<omp_executor> : std::true_type {};
+#endif
+
+template <typename Blocking = blocking_t::always_t,
+          typename ProtoAllocator = std::allocator<void>>
+struct omp_executor {
+  using shape_type = std::size_t;
+
+  using index_type = struct {
+    shape_type max;
+    int idx;
+  };
+
+  std::size_t max_threads = 0;
+
+  omp_executor(): omp_executor(0) {}
+
+  omp_executor(std::size_t max_threads): max_threads(max_threads) {
+#ifdef _OPENMP
+    set_max_threads(max_threads);
+#endif
+  }
+
+  bool set_max_threads(std::size_t max_threads) {
+#ifdef _OPENMP
+    if (!max_threads) this->max_threads = omp_get_max_threads();
+    else this->max_threads = max_threads;
+    return true;
+#endif
+    return false;
+  }
+
+  template <typename Executor, InstanceOf<Executor, omp_executor> = 0>
+  friend bool operator==(const omp_executor& lhs,
+                         const Executor& rhs) noexcept {
+    pcl::utils::ignore(lhs, rhs);
+    return std::is_same<omp_executor, Executor>::value;
+    pcl::utils::ignore(lhs, rhs);
+  }
+
+  template <typename Executor, InstanceOf<Executor, omp_executor> = 0>
+  friend bool operator!=(const omp_executor& lhs,
+                         const Executor& rhs) noexcept {
+    return !operator==(lhs, rhs);
+  }
+
+  template <typename F>
+  void execute(F&& f) const {
+    static_assert(is_executor_available_v<omp_executor>, "OpenMP executor unavailable");
+    f();
+  }
```

Comment:
Pinging @kunaltyagi for opinions

I am not very much in favor of this. Since the initial version of the proposal mentions that `execute` creates a single execution agent and invokes the callable once.

Adding a parallel section which creates a number of execution agents would be violating that definition. 

Currently, `execute` is the same for OMP and inline but that is because of OMP's limited flexibility. 

A better example to differentiate `execute` from `bulk_execute` would be a thread pool executor.
Eg: Let's say we create a thread pool executor with 4 threads.
When we call `execute` on this we are essentially saying invoke the callable on any one of the threads once. This differs from `inline_executor` in some ways, the callable is invoked on one of the threads from the thread pool and not the main thread. Second, if the thread pool executor supports non-blocking behavior then the callable will be invoked in a non blocking manner, whereas inline executor's `execute` can never be non-blocking.
Coming to `bulk_execute`, for thread pool executor it means to creation n execution agents and each agent invoke the callable and pass the agent's index to the callable. Now the mapping between the executor's agents and actual threads is something that can be customized through properties of executors or left to the executor itself.

So by putting `#pragma omp parallel num_threads(max_threads)` in `execute` we are asking it to create multiple execution agents to run the callable on which is the function of `bulk_execute`. We can consider OMP as a specialized case of thread pool where the compiler creates and invokes the callable the specified number of times and then splits the iterations of the loop based on the index of the execution agent passed to it, essentially what `bulk_execute` does.

Reply:
We've tried to stick close to the proposal and let's not throw it away. A single `execute` would have been great for us, but moving away from the proposal has the chance that our implementation will become incompatible with other executor implementations

Reply:
sounds arguments from both. thank you